### ISSUE TITLE
(Auto)Complete features

### ISSUE THREAD

**Issue by MasterMedo** [dense-analysis/ale#3181]
I come from a background of using almost stock *vim/nvim* for 2 years straight, but prior to that I used various IDEs like *Visual Studio* and *IntelliJ*. Kinda still retaining that muscle memory for autocomplete I wanted to try to match the exact behavior of what I remember it being. Upon inspection of the options available for the popup menu I decided I wanted to use the up/down behavior of **state 2** (a match is never selected until I press confirm) so I can scroll through the match list while still being able to type characters if the list is too long. It would be nice if there was a couple of variables in ALE that I could've/can set to get features I want. I read through the FAQ and all the variables in ALEInfo but didn't find these. Here are the features I wanted and how I managed to get most of them:

1. **When opening the popup menu I want the first match to be selected.**
*You get this behavior by clicking `<down>` after opening the complete menu. After an extensive google search + vim-help I found [this](https://github.com/Simpleyyt/.vim/blob/master/autoload/omni/cpp/maycomplete.vim). And solved it by creating a variable that will check if the menu had opened when `<c-x><c-o>` is executed:*
```vim
    let select_first = "\<c-x>\<c-o>"
      \ ."\<C-R>=pumvisible()?\"\\<down>\":\"\"\<cr>"
      \ ."\<cmd>echo\<cr>"

    inoremap <expr> . "." . select_first
```

2. **I want the `tab` to activate the popup menu, confirm an entry if the menu is open, but otherwise act as a normal `tab` key.**
*Luckily when researching ale a few days ago, completely by accident I stumbled upon [this .vimrc](https://github.com/viniciusgerevini/dotfiles/blob/master/vim/.vimrc#L339) where there was a command to do just that.*
```vim
    function! s:check_back_space() abort
      let col = col('.') - 1
      return !col || getline('.')[col - 1]  =~# '\s'
    endfunction

    inoremap <expr> <tab> pumvisible() ? "\<c-y>"  :
             \ <sid>check_back_space() ? "\<tab>"  : select_first
```

3. **When selecting an item from the popup menu I don't want it to ever insert trailing dots.**
*I presume this has nothing to do with ALE but rather with* _**pyls**_. *For example when selecting a match for the package `json` a trailing dot is automatically inserted. I managed to solve it using an `autocmd`. And I am aware that I might need to check if the key 'word' is in the `v:completed_item` dictionary.*
```vim
  autocmd CompleteDone <buffer> if v:completed_item.word =~ '\.$'
                             \| call feedkeys("\<bs>")
                             \| endif
```

4. **Similar problem to `3.`, I want the matching parenthesis to be inserted along with the match**
*So when I autocomplete a function e.g. `json.detect_encoding(` I want to get `json.detect_encoding()` with my cursor being placed inside the parenthesis. This I found [here](https://vi.stackexchange.com/questions/10650/add-closing-parenthesis-when-completing-function-omnicomplete).*
```vim
  autocmd CompleteDone <buffer> if v:completed_item.word =~ '($'
                             \| call feedkeys(")\<Left>", 'in')
                             \| endif
```

It would be nice if we had some variables within ALE like `g:ale_completion_select_first`, `g:ale_completion_use_tab`, `g:ale_completion_trim_chars` and `g:ale_completion_append_chars`.

**Comment by ipkiss42** [comment]
I'm not an ALE contributor, but it seems to me that nothing here is ALE-specific.

For (1), I suspect you have `noselect` in `'completeopt'`, see `:h completeopt`. A dedicated mapping (or ALE option) shouldn't be needed.

For (2), it is simply a preference. Personally I have something similar, but when not in the popup menu I use `<Tab>` to expand a snippet or jump to the next placeholder, and finally fall back to a regular `<Tab>`. Other users likely have different preferences (e.g. using a different snippet plugin), so IMHO it doesn't make sense to add an option for every single one of them. That's what .vimrc is for :)

As you said, (3) sounds specific to `pyls` (you could confirm it using a different LSP client and/or server). I don't think ALE inspects the completions provided by the server (and I don't believe it should), so any customization seems more appropriate in the server or in your own config.

Same for (4): it is your preference, but other users might want the matching parenthesis only for some filetypes, or only if there is no argument (and thus without moving the cursor to the left), etc. An option saving three lines of .vimrc, that needs to be tested, maintained, and may not be useful to many users, sounds like useless complexity to me.

Again, this is only my personal opinion. ALE authors might disagree.

**Comment by MasterMedo** [comment]
> For (1), I suspect you have noselect in 'completeopt', see :h completeopt. A dedicated mapping (or ALE option) shouldn't be needed.

That is why I specifically stated and bolded that I want to start in the **2. state** of completion, I don't want the match to be written out like when using `<c-n>`/`<c-p>`, I want to be able to type letters to find the match **and** be able to go up and down the list using `<down>`/`<up>`. So to achieve that behavior you must either have `noselect` or `longest` added and I do have `longest` (I host my vimrc on github so you can check out the rest).

ALE provides built in completion and already offers some variables to customize it. I'm not sure how well the Language Server Protocol is defined, does it specify should dots/parenthesis be added after functions/class objects/variables. If that is the case then the program executing the completion or vim itself is exactly the place where these options should be available. The way I solved these issues is in a rather hacky manner.

**Comment by ipkiss42** [comment]
> That is why I specifically stated and bolded that I want to start in the **2. state** of completion, I don't want the match to be written out like when using `<c-n>`/`<c-p>`, I want to be able to type letters to find the match **and** be able to go up and down the list using `<down>`/`<up>`. So to achieve that behavior you must either have `noselect` or `longest` added and I do have `longest` (I host my vimrc on github so you can check out the rest).

It sounds like you need the `noinsert` value in `completeopt`, still without `noselect`.

> ALE provides built in completion and already offers some variables to customize it. I'm not sure how well the Language Server Protocol is defined, does it specify should dots/parenthesis be added after functions/class objects/variables. If that is the case then the program executing the completion or vim itself is exactly the place where these options should be available. The way I solved these issues is in a rather hacky manner.

Why is it hacky?

As I explained above, having config options to handle every single use case does not scale well given the number of possible use cases (c.f. a few examples in my previous post).

**Comment by MasterMedo** [comment]
> It sounds like you need the `noinsert` value in `completeopt`, still without `noselect`.

Yeah, I know that works but only without `longest`.

> Why is it hacky?

Modifying the result instead of redefining the process is hacky in and of itself.
The 1) option should probably be fixed on vim's side. The documentation speaks of no correlation between `longest` and `noselect`.
The 2) I completely agree with you, should stay in a vimrc file. I'll look into writing a function for that one.
The 3) and 4) are trivially hacky, fixing behavior instead of redefining it. And although I agree there are user preferences as the ones you mentioned there possibly aren't so many for common languages. An option like `trim_characters` which would be an array of characters for you to define seems reasonable for a program that does completion, I don't believe that should go in every single language server out there. Complexity does increase but providing common preferences isn't radical at all, even having 10 options like the ones we listed seems very reasonable to me. Vim comes with a lot of options and having a plugin increases complexity enough, there is no need to have a plugin for a plugin (even a personal one).

**Comment by MasterMedo** [comment]
I tried a couple of other completion plugins. I see `ale` is far from having a full completion suite like `coc-nvim`, which has integrated solutions for some of the features I proposed, be it as options or sub-plugins so to call them. Also, I've read that using vims built in `omnifunc` is blocking but having a separate `popupmenu` like `coc` has can be asynchronous, although I have to say A.L.E's completion was noticeably faster.

Indeed 3) was due to `pyls`. `jedi` for python doesn't even put parenthesis for functions.
I also see there exist dedicated language servers like `tabnine` that support features like 3) and 4) and the ones you mentioned (having a placeholder for you to jump to for each function argument and closing the function parenthesis, leaving your cursor after it).
Interestingly enough language servers can provide text snippets!

I'll close this cause it seems out of scope for ALE.
